### PR TITLE
object creation tests

### LINKED ISSUE #1600: Finish object creation tests

### PR DESCRIPTION
**Pull request by segiddins**
Closes #1600.

@alazier I started these, but I still need to do the `add` and `addOrUpdate` tests. Opening a PR in case you wanted to augment them

### CODE REVIEW

**Comment by alazier** [comment]
do we need a rebase or something?

**Comment by segiddins** [comment]
Apparently :(

**Comment by segiddins** [comment]
Rebased again

File: RealmSwift/Tests/ObjectCreationTests.swift [hunk 1]
```diff
@@ -312,7 +397,7 @@ class ObjectCreationTests: TestCase {
             case .String:   return ["b"]
             case .Data:     return ["b".dataUsingEncoding(NSUTF8StringEncoding, allowLossyConversion: false)! as NSData]
             case .Date:     return [NSDate(timeIntervalSince1970: 2) as AnyObject]
-            case .Object:   return [[true], ["boolCol": true], SwiftBoolObject(object: [true])]
+            case .Object:   return [[true], ["boolCol": true], SwiftBoolObject(object: [true]), persistedObject]
             case .Array:    return [[[true], [false]], [["boolCol": true], ["boolCol": false]], [SwiftBoolObject(object: [true]), SwiftBoolObject(object: [false])]]
```
Comment (alazier) [inline RealmSwift/Tests/ObjectCreationTests.swift]: We should have an array of persisted objects

File: RealmSwift/Tests/ObjectCreationTests.swift [hunk 2]
```diff
@@ -237,15 +237,74 @@ class ObjectCreationTests: TestCase {
...
+        let standalone = SwiftPrimaryStringObject(object: ["11", 11])
+        Realm().beginWrite()
+        let objectWithNestedObjects = Realm().create(SwiftLinkToPrimaryStringObject.self, value: ["11", ["11", 11], [standalone]])
+        Realm().commitWrite()
+
+        let stringObjects = Realm().objects(SwiftPrimaryStringObject)
+        XCTAssertEqual(stringObjects.count, 1)
+        let persistedObject = stringObjects.first!
+
+        XCTAssertNotEqual(standalone, persistedObject)
```
Comment (alazier) [inline RealmSwift/Tests/ObjectCreationTests.swift]: comment here would be useful - we test that standalone object is copied into realm

File: RealmSwift/Tests/ObjectCreationTests.swift [hunk 3]
```diff
@@ -237,15 +237,74 @@ class ObjectCreationTests: TestCase {
...
+        XCTAssertEqual(objectWithNestedObjects.object!, persistedObject)
+        XCTAssertEqual(objectWithNestedObjects.objects.first!, persistedObject)
     }
 
     func testUpdateWithNestedObjects() {
-    }
+        let standalone = SwiftPrimaryStringObject(object: ["11", 11])
```
Comment (alazier) [inline RealmSwift/Tests/ObjectCreationTests.swift]: might be more clear to use the string "primary" for the primary key value.

Comment (alazier) [inline RealmSwift/Tests/ObjectCreationTests.swift]: that goes for all of these tests.

File: RealmSwift/Tests/ObjectCreationTests.swift [hunk 4]
```diff
@@ -237,15 +237,74 @@ class ObjectCreationTests: TestCase {
...
+        XCTAssertNotEqual(standalone, persistedObject) // standalone object should be copied into the realm, not added directly
+        XCTAssertEqual(objectWithNestedObjects.object!, persistedObject)
+        XCTAssertEqual(objectWithNestedObjects.objects.first!, persistedObject)
     }
 
     func testUpdateWithNestedObjects() {
-    }
+        let standalone = SwiftPrimaryStringObject(object: ["primary", 11])
+        Realm().beginWrite()
+        let object = Realm().create(SwiftLinkToPrimaryStringObject.self, value: ["otherPrimary", standalone, [["primary", 12]]], update: true)
```
Comment (alazier) [inline RealmSwift/Tests/ObjectCreationTests.swift]: should test with update:false as well

File: RealmSwift/Tests/ObjectCreationTests.swift [hunk 5]
```diff
@@ -237,15 +237,74 @@ class ObjectCreationTests: TestCase {
...
+        realmWithTestPath().beginWrite()
+        let otherRealmObject = realmWithTestPath().create(SwiftLinkToPrimaryStringObject.self, value: ["primary", NSNull(), [["2", 2], ["4", 4]]])
+        realmWithTestPath().commitWrite()
+
+        Realm().beginWrite()
+        Realm().create(SwiftLinkToPrimaryStringObject.self, value: ["primary", ["10", 10], [["11", 11]]])
+        let object = Realm().create(SwiftLinkToPrimaryStringObject.self, value: otherRealmObject, update: true)
+        Realm().commitWrite()
+
+        XCTAssertNotEqual(otherRealmObject, object) // the object from the other realm should be copied into this realm
```
Comment (alazier) [inline RealmSwift/Tests/ObjectCreationTests.swift]: please test count of objects in the target realm. Also test all object values are what they are expected to be.

**Comment by segiddins** [comment]
@alazier could you take another look?

**Pull request merged**